File: button_diode_switch/src/button_diode_switch.py
import RPi.GPIO as GPIO
import time

ledPin = 11
buttonPin = 12
light = False
button_pressed = False

def setup():
	print('Program is starting...')
	GPIO.setmode(GPIO.BOARD)
	GPIO.setup(ledPin, GPIO.OUT)
	GPIO.setup(buttonPin, GPIO.IN, pull_up_down=GPIO.PUD_UP)

def loop():
	global light
	global button_pressed
	while True:
		if GPIO.input(buttonPin) == GPIO.LOW:
			time.sleep(0.1)
			if light == False:
				GPIO.output(ledPin, GPIO.HIGH)
				print('led on...')
			else:
				GPIO.output(ledPin, GPIO.LOW)
				print('led off...')
			button_pressed = True
		elif button_pressed == True:
			light = not light
			button_pressed = False

def destroy():
	GPIO.output(ledPin, GPIO.LOW)
	GPIO.cleanup()
	
if __name__ == '__main__':
	setup()
	try:
		loop()
	except KeyboardInterrupt:
		destroy()

File: button_diode_switch/src/button_diode_switch.c
#include <stdio.h>
#include <wiringPi.h>

#define ledPin 0
#define buttonPin 1

int main(void) {
  int light, button_pressed;
  light = 0;
  button_pressed = 0;
  if (wiringPiSetup() == -1) {
    printf("Setup wiringPi failed!\n");
    return 1;
  }

  pinMode(ledPin, OUTPUT);
  pinMode(buttonPin, INPUT);

  pullUpDnControl(buttonPin, PUD_UP);
  digitalWrite(ledPin, LOW);

  while (1) {
    if (digitalRead(buttonPin) == LOW) { // button pressed
      delay(100);
      if (light == 0) { // if light off
        digitalWrite(ledPin, HIGH);
        printf("led on...\n");
      } else { // if light on
        digitalWrite(ledPin, LOW);
        printf("...led off\n");
      }
      button_pressed = 1;
    } else if (button_pressed == 1) { // button released
      if (light == 0) {
        light = 1;
      } else {
        light = 0;
      }
      button_pressed = 0;
    }
  }

  return 0;
}
